R/send_sparql.R

```diff
@@ -12,38 +12,70 @@
 #'} ORDER BY ?itemLabel
 #''
 #'send_sparql(query=metro_query)
-#' @import SPARQL
 send_sparql=function(query,endpoint="Wikidata"){
   endpoint=tolower(endpoint)
 
   # if endpoint wikidata, use WikidataQueryServiceR::query_wikidata()
   if(endpoint=="wikidata"){
-    tib <- purrr::quietly(WikidataQueryServiceR::query_wikidata)(query)$result
+    return(purrr::quietly(WikidataQueryServiceR::query_wikidata)(query)$result)
   }
-  # else, use SPARQL::SPARQL()
+  # else, use httr2
+
   # if endpoint passed as name, get url
-  if(endpoint %in% (usual_endpoints %>%
-                    dplyr::filter(.data$name!="wikidata") %>%
-                    dplyr::pull(.data$name))){
-    url=usual_endpoints %>%
-      dplyr::filter(.data$name == {{ endpoint }}) %>%
-      dplyr::select(.data$url) %>%
-      dplyr::pull()
-    tib <- SPARQL::SPARQL(url=url,
-                          query=query,
-                          curl_args=list(useragent='glitter R package (https://github.com/lvaudor/glitter)'))
-    tib <- tib$results
-  }
-  # else, endpoint must be passed as url
-  if(!(endpoint %in% usual_endpoints$name)){
-    tib <- SPARQL::SPARQL(url=endpoint,
-                          query=query,
-                          curl_args=list(useragent='User Agent Example'))
-    tib <- tib$results
+  usual_endpoint_info = usual_endpoints %>%
+                    dplyr::filter(.data$name == endpoint)
+  url = if (nrow(usual_endpoint_info) > 0) {
+    dplyr::pull(usual_endpoint_info, .data$url)
+  } else {
+    endpoint
   }
-  # if returned table has 0 rows replace it with NULL (easier to bind with other results)
-  if(nrow(tib)==0){tib=NULL}
-  return(tib)
+
+
+    resp = httr2::request(url) %>%
+    httr2::req_url_query(query = query) %>%
+    httr2::req_method("POST") %>%
+    httr2::req_headers(Accept = "application/sparql-results+json") %>%
+    httr2::req_user_agent("glitter R package (https://github.com/lvaudor/glitter)") %>%
+    httr2::req_retry(max_tries = 3, max_seconds = 120) %>%
+    httr2::req_perform()
+
+    httr2::resp_check_status(resp)
+
+    if (httr2::resp_content_type(resp) != "application/sparql-results+json") {
+      rlang::abort("Not right response type") #TODO:better message, more flexibility
+    }
+
+    content = httr2::resp_body_json(resp)
+
+    # Adapted from https://github.com/wikimedia/WikidataQueryServiceR/blob/accff89a06ad4ac4af1bef369f589175c92837b6/R/query.R#L56
+    if (length(content$results$bindings) > 0) {
+        data_frame <- purrr::map_dfr(content$results$bindings, function(binding) {
+          return(purrr::map_chr(binding, ~ .x$value))
+        })
+        datetime_columns <- purrr::map_lgl(content$results$bindings[[1]], function(binding) {
+          if ("type" %in% names(binding)) {
+            return(binding[["type"]] == "http://www.w3.org/2001/XMLSchema#dateTime")
```

Comment:
to be tested with queries where we know there will be a datatime, and to be generalized to other special types.